contracts/token/erc721/ImmutableERC721RoyaltyEnforced.sol

```diff
@@ -0,0 +1,94 @@
+//SPDX-License-Identifier: Apache 2.0
+pragma solidity ^0.8.0;
+
+// Token
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+// Allowlist Registry
+import "../../royalty-enforcement/IRoyaltyAllowlist.sol";
+
+/*
+    ImmutableERC721RoyaltyEnforced is an abstract contract extending the ERC721 contract to enforce royalties.
+    The contract has support to specify a whitelist registry and overrides transfers and approvals to enforce allow list validation.
+*/
+
+abstract contract ImmutableERC721RoyaltyEnforced is
+    ERC721
+{
+    ///     =====     Errors         =====
+
+    /// @dev Error thrown when calling address is not Allowlisted
+    error CallerNotInAllowlist(address caller);
+
+    /// @dev Error thrown when approve target is not Allowlisted
+    error ApproveTargetNotInAllowlist(address target);
+
+    ///     =====     Events         =====
+
+    /// @dev Emitted whenever the transfer Allowlist registry is updated
+    event RoyaltytAllowlistRegistryUpdated(address oldRegistry, address newRegistry);
+
+    ///     =====   State Variables  =====
+
+    /// @dev Interface that implements the `IRoyaltyAllowlist` interface
+    IRoyaltyAllowlist public royaltyAllowlist;
+
+    /// @dev Allows admin to set or update the royalty Allowlist registry
+    function setRoyaltyAllowlistRegistry(address _royaltyAllowlist) public virtual {
+        require(IERC165(_royaltyAllowlist).supportsInterface(type(IRoyaltyAllowlist).interfaceId), "contract does not implement IRoyaltyAllowlist");
+
+        emit RoyaltytAllowlistRegistryUpdated(address(royaltyAllowlist), _royaltyAllowlist);
+        royaltyAllowlist = IRoyaltyAllowlist(_royaltyAllowlist);
+    }
+
+    /// @dev Override of setApprovalForAll from {ERC721}, with added Allowlist approval validation
+    function setApprovalForAll(address operator, bool approved) public virtual override {
+        _validateApproval(operator);
+        super.setApprovalForAll(operator, approved);
+    }
+
+    /// @dev Override of approve from {ERC721}, with added Allowlist approval validation
+    function approve(address to, uint256 tokenId) public virtual override {
+        _validateApproval(to);
+        super.approve(to, tokenId);
+    }
+
+    /// @dev Internal function to validate whether approval targets are Allowlisted or EOA
+    function _validateApproval(address targetApproval) internal view {
+        // Only check if the registry is set
+        if(address(royaltyAllowlist).code.length > 0) {
+             // Check for:
+            // 1. approval target is an EOA
+            // 2. approval target address is Allowlisted or target address bytecode is Allowlisted
+            if (targetApproval.code.length == 0 || royaltyAllowlist.isAllowlisted(targetApproval)){
```

Comment:
interesting choice to do the EOA check in the royalty enforcing contract as opposed to the `isAllowlisted` method - have you considered moving this check into the allowlist itself for completeness?

Reply:
I kept it here as really the main purpose of the `royaltyAllowlist` is a registry of publicly authorized operators and validating whether a supplied address is one of those operators. Being strict about semantics here, EOA's are not one of these authorized operators but rather an entity that does not concern themselves with royalty enforcement, i.e. the transfer behaviour to and from EOAs has not changed with the inclusion of `royaltyAllowlist`.